bokehjs/src/coffee/api/charts.coffee

```diff
@@ -0,0 +1,138 @@
+_ = require("underscore")
+$ = require("jquery")
+sprintf = require("sprintf")
+{Document} = require("../document")
+embed = require("../embed")
+models = require("./models")
+palettes = require("../palettes/palettes")
+
+sum = (array) ->
+  return array.reduce(((a, b) => a + b), 0)
+
+cumsum = (array) ->
+  result = []
+  array.reduce(((a, b, i) -> result[i] = a + b), 0)
+  return result
+
+num2hexcolor = (num) -> sprintf("#%06x", num)
+
+pie = (data, opts) ->
+  labels = _.clone(data.labels)
+  values = _.clone(data.values)
+
+  i = 0
+  while i < values.length
+    if values[i] <= 0
+      labels.splice(i, 1)
+      values.splice(i, 1)
```

Comment:
This is `O(n^2)`. It can be done in linear time with a normal for-loop which filters the values pair-wise. You can also achieve the shallow-copy at the same time, making the whole lot more efficient.